csrc/id_model/id_graph.h

```diff
@@ -0,0 +1,255 @@
+
+  // Convert iter domain to its IdGroup, assert that it exists.
+  IdGroup toGroup(IterDomain* id) const;
+
+  // Convert unique vector of expressions to unique vector of its groups
+  ExprGroups toGroups(const VectorOfUniqueEntries<Expr*>& exprs) const;
+
+  // Convert unique vector of IterDomain to unique vector of its groups
+  IdGroups toGroups(const VectorOfUniqueEntries<IterDomain*>& ids) const;
+
+  // Return output/input iter domain groups of provided expr
+  std::vector<IdGroup> outputGroups(ExprGroup expr) const;
+  std::vector<IdGroup> inputGroups(ExprGroup expr) const;
+
+  // Traverses uses of the IdGroups in 'of' and returns all ExprGroups
+  // that have a use in their definition of provided of IdGroups.
+  ExprGroups allUsesOf(const IdGroups& of) const;
+
+  // Traverses definitions of the IdGroups in 'of' and returns all ExprGroups
+  // used in this history of defining the 'of' IdGroups.
+  ExprGroups allDefinitionsOf(const IdGroups& of) const;
+
+  // Return sorted expressions to go from the provided IterDomains in from to
+  // the provided IterDomains in to with provided mode. Minimal expressions to
+  // get from 'from' to 'to' returned.
+  ExprGroups getExprsBetween(const IdGroups& from, const IdGroups& to) const;
+
+  // Supports one to many mappings, uses the disjoint sets of the provided mode
+  // to produce mappings between from and to. If multiple IterDomains in to map
+  // to a single iter domain in from, the order of the IterDomains in value of
+  // the map is preserved to be the order provided in to.
+  std::unordered_map<IterDomain*, VectorOfUniqueEntries<IterDomain*>>
+  buildMapBetween(
+      const std::vector<IterDomain*>& from,
+      const std::vector<IterDomain*>& to) const;
+
+  // Alias of the above on unique vector entries
+  std::unordered_map<IterDomain*, VectorOfUniqueEntries<IterDomain*>>
+  buildMapBetween(
+      const VectorOfUniqueEntries<IterDomain*>& from,
+      const VectorOfUniqueEntries<IterDomain*>& to) const;
+
+  //! Returns
+  //!   (1) The expressions associated with the definitions of the provided
+  //!     IterDomain group in the provided mapping mode (if it exists).
+  //!   (2) If there is a definitions entry of the provided IterDomain group in
+  //!     the provided mapping mode.
+  //! First entry in the returned pair is a vector of vector of expressions. The
+  //! inner vector is proven to be equivalent based on the provided mode. The
+  //! outer vector are expression groups that are not equivalent based on the
+  //! provided mode, but produce one of the IterDomains within the same disjoint
+  //! Iter Domain set based on the provided mode.
+  //! TODO: Change name to start with get
+  std::pair<ExprGroups, bool> iterDomainGroupDefinitions(
+      IdGroup id_group) const;
+
+  //! Same as iterDomainGroupDefinitions but for uses instead of definitions
+  //! TODO: Change name to start with get
+  std::pair<ExprGroups, bool> iterDomainGroupUses(IdGroup id_group) const;
```

Comment:
Same here, looks like we build out a special IR node that represent a collection of other IR nodes. Would it be possible to achieve this by just subclassing `Val` and `Expr`, and maybe also override `inputs()`, `outputs()` if needed?

Reply:
I think it would require a full container. I.E. start with subclassing IrContainer/Fusion/or Kernel. However, I'm still not 100% sure how we want to interact with this system so hesitant to go down that route right now.

Reply:
Should still cleanup the pair usage here.